src/persist-client/src/cli/inspect.rs

```diff
@@ -630,6 +643,204 @@ pub async fn unreferenced_blobs(args: &StateArgs) -> Result<impl serde::Serializ
+                }
+                Err(err) => {
+                    eprintln!("error parsing blob: {}", err);
+                }
+                _ => (),
+            }
+        })
+        .await?;
+
+        Ok(shards)
+    }
+
+    // Open a handle to S3.
+    let blob = make_blob(&cfg, blob_uri, NO_COMMIT, Arc::clone(&metrics))
+        .await
+        .context("making blob")?;
+    tracing::info!(?blob_uri, "opened blob");
+
+    // List all shards.
+    let shards = mz_ore::retry::Retry::default()
+        .max_tries(5)
+        .retry_async(|_| {
+            let blob = Arc::clone(&blob);
+            async move { list_shards(Arc::clone(&blob)).await }
+        })
+        .await?;
+
+    let num_parts: usize = shards.values().map(|parts| parts.len()).sum();
+    tracing::info!(
+        ?blob_uri,
+        num_shards = shards.len(),
+        num_parts,
+        "listed blob"
+    );
+
+    let mut report = BlobReport::default();
+    for (shard_id, partial_keys) in shards {
+        // Drives all of the GET futures concurrently.
+        //
+        // Note(parkmycar): We should limit the number of blobs we fetch in parallel to reduce
+        // memory usage.
+        let mut blob_stream: FuturesUnordered<_> = partial_keys
+            .iter()
+            .map(|partial_key| {
+                let blob = Arc::clone(&blob);
+                let key = partial_key.complete(&shard_id);
+                mz_ore::retry::Retry::default().retry_async(move |_| {
+                    let blob = Arc::clone(&blob);
+                    let key = key.clone();
+                    async move { blob.get(&key).await }
+                })
+            })
+            .collect();
+
+        // Updates the 'persist_use_arrow_rs_library' feature flag.
+        let update_feature_flag = |val: &str| {
+            let mut updates = mz_dyncfg::ConfigUpdates::default();
+            let val = mz_dyncfg::ConfigVal::String(val.to_string());
+            updates.add_dynamic("persist_use_arrow_rs_library", val);
+            updates.apply(metrics.columnar.cfg());
+        };
+
+        while let Some(blob_result) = blob_stream.next().await {
+            let Some(blob) = blob_result? else {
+                report.missing_blobs += 1;
+                continue;
+            };
+
+            // Decode with both `arrow2` and `arrow-rs`.
+            let arrow2_records: BlobTraceBatchPart<i64> = {
+                update_feature_flag("off");
+                decode_trace_parquet(blob.clone(), &metrics.columnar)?
+            };
+            let arrow_rs_records: BlobTraceBatchPart<i64> = {
+                update_feature_flag("read_and_write");
+                decode_trace_parquet(blob.clone(), &metrics.columnar)?
+            };
+
+            // Make sure the records match.
+            if arrow2_records != arrow_rs_records {
+                anyhow::bail!("arrow2 and arrow-rs return different values, {blob_uri:?}");
```

Comment:
We want to include the blob key in this too, no?

Reply:
updated!